lib/api/ldp/ldp-request.js

```diff
@@ -0,0 +1,145 @@
+'use strict'
+
+const HttpError = require('standard-http-error')
+
+class LdpRequest {
+  /**
+   * @param options {object}
+   *
+   * @param options.host {SolidHost} Host / config object
+   *
+   * @param options.target {URL} Parsed full request url
+   *
+   * @param options.req {HttpRequest} Save the Express req object, useful
+   *   to access headers/preferences, writeStream for body, etc.
+   *
+   * @param [options.authn=null] {object} Authentication object, contains WebID
+   *   string, as well as any bearer credentials/tokens. Needed for
+   *   authenticated fetch (of remote group ACLs, of Copy resources, etc).
+   *   Null if request is not authenticated.
+   *
+   * @param [options.parsedBody] {?} Request body, parsed when appropriate
+   *   (when content-type is parse-able)
+   */
+  constructor (options) {
+    this.host = options.host
+    this.target = options.target
+    this.req = options.req
+    this.authn = options.authn
+    this.parsedBody = options.parsedBody
+  }
+
+  static from ({req, host}) {
+    let request
+
+    switch (req.method) {
+      case 'head':
+        /**
+         * Checks to see if target exists
+         */
+        request = LdpHeadRequest.from({req, host})
```

Comment:
Where do these come from? (I assume to be implemented still?)

Also, I wonder whether it's necessary to do the subclassing at this level, but it is an option. That said, we there's probably some common parsing to be done first.

Reply:
Well, actual operation-level subclasses :smile:

These are request-level subclasses.
I would expect to have one `LdpRequest` which _has_ an operation, and operations can then have different subclasses.

Not much difference I guess, slightly different way of modeling.

Reply:
What do they say about naming being the hardest problem in CS? :)
I think it would be clearer if these subclasses are named e.g. `LdpGetOperation`, instead of `LdpGetRequest` -- cause they're modeling operations, not http requests. We leave request land in that `switch()` statement.